src/webextension/background/browser-action.js

```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { openView } from "./views";
+
+let listener;
+let popup;
+
+function installPopup(path) {
+  popup = browser.extension.getURL(path);
+  browser.browserAction.setPopup({
+    popup,
+  });
+}
+function uninstallPopup() {
+  if (popup) {
+    browser.browserAction.setPopup({ popup: "" });
+  }
+  popup = null;
+}
+
+function installListener(name) {
+  listener = () => { openView(name); };
+  browser.browserAction.onClicked.addListener(listener);
+}
+function uninstallListener() {
+  if (listener) {
+    browser.browserAction.onClicked.removeListener(listener);
+  }
+  listener = null;
+}
+
+export default async function updateBrowserAction(ds) {
+  // clear listener
+  // XXXX: be more efficient with this?
+  uninstallListener();
+  uninstallPopup();
+
+  if (!ds.initialized) {
+    // setup first-run popup
+    console.log("setup first run click handler");
+    return installListener("firstrun");
+  } else if (ds.locked) {
+    // setup unlock popup
+    console.log("setup unlock popup");
+    return installPopup("popup/unlock.html");
+  }
```

Comment:
Are we sure we want to do unlocking as a popup? What should happen when the user opens a management tab and then the datastore locks later on? I know this isn't something we're doing right now, but I want to make sure we're not painting ourselves into a corner here.

If we do want to go this way, we could probably simplify things by *always* having `onClick` set and picking what tab to open based on whether the datastore is initialized. Per MDN:

> [`browserAction.onClicked` is] Fired when a browser action icon is clicked. This event will not fire if the browser action has a popup.

Reply:
I think unlock is going to be a popup always.  For an open management tab, there's a couple of options I can see:
* close the management tab; this seems the least desirable though
* blank everything, and instruct the user to click the popup to unlock

Either way, we'll have to deal with pending changes somehow...

Reply:
> f we do want to go the way you have it, we could probably simplify things by always having onClick set and picking what tab to open based on whether the datastore is initialized. Per MDN:

>>    [browserAction.onClicked is] Fired when a browser action icon is clicked. This event will not fire if the browser action has a popup.

That is a good catch, but this code is partly clearing the click handler each time in order to simplify the switch between FTUX and edit/manage.